(wip) Adding first part of spatial segregation measures

renanxcortes opened the pull request:
Serge, this pull request adds several more functions of the segregation module that take in consideration the spatial context. The new indices of the functions are (alphabetically ordered):

- Absolute Centralization
- Absolute Concentration
- Boundary Spatial Dissimilarity
- Delta
- Perimeter Area Ratio Spatial Dissimilarity
- Relative Centralization
- Relative Clustering
- Relative Concentration
- Spatial Dissimilarity
- Spatial Exposure
- Spatial Isolation
- Spatial Proximity
- Spatial Proximity Profile


Also, I added two inference function that wraps the segregation measures. The first is 'infer_segregation.py' that has the Infer_Segregation class for a single measure and 'compare_segregation.py' that has the Compare_Segregation class to perform inference for two measures.

Some aspects that it's good to mention:

- I kept the example with raw census data set instead of using the harmonized osnap for two reasons: firstly, since it uses the sample data and all the values and examples would have to change (in some cases the difference was not negligible) and, secondly, some literature say that since the dissimilarity measures is upward biased working with one year with sample and other year with the full count could affect the comparisons between two periods (see Reardon (2018) "Has Income Segregation Really Increased? Bias and Bias Correction in Sample-Based Segregation Estimates"). However, I changed the examples following the step-by-step procedure of osnap (I linked to osnap) so the user is able to follows it and save the file as filepath = '~/data/std_2010_fullcount.csv'.

- I removed the data from the GitHub repo so we won't have any problem with that. However, since the new measures work with spatial context, I created another repo that contains the .json of the maps to geopandas read them. These .json will not have any problem since they're not sensitive, right?

- The most trick functions were the boundary spatial dissimilarity and the perimeter/area ratio dissimilarity from Wong (1993). Levi helped me to create a function that return the length border of a given geopandas (it's present in the "util.py" script, but I think it can be polished it and added in libpysal). It was not easy to reproduce the values of the original paper and the comparison of the seg R package. In R, I couldn't compare with any personalized dataset since it relies on the installation of older version of GRASS and I had the most recent one which was not compatible. However, I could compare the values with the toy examples "lattices" of the Wong (1993) paper. The notebook called "Comparisons_Spatial_Dissimilarities_with_seg_R_package" kind of comprises roughly my comparisons. Moreover, the comparison of the R package and the measures of the papers is here: https://sites.google.com/site/hongseongyun/seg

- Some useful notebook that I was using/working to study the measures and compare them are also presented in this Pull Request. They're not as organized as the actual functions are.

Best,
Renan

renanxcortes committed:
updating the example with user download data in the docstring

renanxcortes committed:
fixing identation in docstring

renanxcortes committed:
identation fixed

renanxcortes committed:
fixing identation

renanxcortes committed:
fixing identation

renanxcortes committed:
fixing the rest of the identation in the docstrings

renanxcortes committed:
fixing identation

renanxcortes committed:
adding spatial dissimilarity function

This is the spatial dissimilarity index from Morril (1991). It the analogous of the DM measure of the R seg package.

renanxcortes committed:
adding condition of the w argument

It has to be a PySAL weights object

renanxcortes committed:
include spatial proximity measure

renanxcortes committed:
change to spatial_prox_index object name in the docstring example

renanxcortes committed:
adding relative clustering measure

renanxcortes committed:
fixing code header

renanxcortes committed:
adding spatial isolation

renanxcortes committed:
adding spatial exposure

renanxcortes committed:
adding the interpretation in the Note docstring

renanxcortes committed:
adding interpretation in the Note in the docstring

renanxcortes committed:
adding Delta measure

renanxcortes committed:
adding relative concentration index

renanxcortes committed:
adding absolute centralization measure

renanxcortes committed:
fixing the acronym index in the example

renanxcortes committed:
adding relative centralization index

renanxcortes committed:
removing not applicable comment

renanxcortes committed:
adding absolute concentration index  and spatial isolation/exposure

renanxcortes committed:
adding absolute concentration

renanxcortes committed:
adding util.py

renanxcortes committed:
add an argument for row standardization (the seg package works with standardized values)

renanxcortes committed:
adding boundary spatial dissimilarity this is the Modified D (boundary length), DW by Hong (2014)

it uses a function originally developed by Levi in the util.py that calculates the border length of contiguity neighbors

renanxcortes committed:
adding notebook that compares some values of the original paper os spatial dissimilarity indexes of Wong (1993)

renanxcortes committed:
optimizing code by referring the raw D to its function

renanxcortes committed:
removing dissimilarity from notebooks repo

renanxcortes committed:
optimizing code to use dissimilarity function already built

renanxcortes committed:
adding boundary spatial dissimilarity index

renanxcortes committed:
removing comment that the neighborhood contiguity matrix is used which is not needed in this index

renanxcortes committed:
currently the pard function that is more similar to the original paper of Wong (1993)

renanxcortes committed:
adding perimeter/area ratio spatial dissimilarity

renanxcortes committed:
removing Hong (2014) from the docstring... since function needs a second look, since some values are not matcing from seg package of R in their  asymetrical lattices examples

renanxcortes committed:
adding the comparions with seg R package

renanxcortes committed:
adding toy example of potential computational opportunity for RIDIR scaling in the inference framework

renanxcortes committed:
add comparison between osnap data and raw census data of 2010

renanxcortes committed:
add Leiv's notebook of the weighted w functions of the shared borders

renanxcortes committed:
adding the spatial proximity profile function

renanxcortes committed:
adding description of SPP

renanxcortes committed:
correcting reference in the docstring

renanxcortes committed:
adding the spatial proximity profile function

renanxcortes committed:
fixing the object name in the plot method in the docstring

renanxcortes committed:
add a wrapper to perform inference on a single segregation measure

ljwolf commented on inequality/boundary_spatial_dissimilarity.py:
```diff
@@ -0,0 +1,149 @@
+
+import numpy as np
+import pandas as pd
+#import libpysal
```
strike

ljwolf commented on inequality/perimeter_area_ratio_spatial_dissimilarity.py:
```diff
@@ -0,0 +1,148 @@
+    total_pop_var : string
+                    The name of variable in data that contains the total population of the unit
+                    
+    std           : boolean
```
call this `standardize`? `std` is pretty vague.

renanxcortes replied:
Yes! I'll change to 'standardize' just like other builtin PySAL functions.

ljwolf replied:
Not as important, but your other variable names are so good and explanatory, this would be better if it matched :smile:

renanxcortes replied:
No problem!

renanxcortes committed:
changing the parameter name to 'standardize'

renanxcortes committed:
changing the parameter name to standardize

renanxcortes committed:
adding core_data public argument

this commit adds a core_data public argument in all functions and adapt everything that rely on each function. For example, if a measure relies on the _dissim function, it will access the value of the dissimilarity with [0] index. In addition, in the definition of the class, I put an auxiliary object (called 'aux') in order to avoid multiple calls of the same function in order to be more efficient.

renanxcortes committed:
this commit puts an extra condition in the systematic and eveness approach for geopandas. Also, it put an 'aux' object the function to call to be more effective.

renanxcortes committed:
adds a two_tailed argument and a statistic attribute

renanxcortes committed:
adds a wrapper function to perform inference between two segregation measures

this commit adds the main inference function to perform inference for two segregation measures. The null_approach can be either random_data or random_geometry which affects differently the iterations. This functions excpects the user to put as input two segregation previously already built in this module. In addition, since some measures rely on different parameters (for example, alpha, beta, b, m, etc.) the user need to put explicitly again the arguments in this function with **kwargs.

renanxcortes committed:
fixing the name of the class to 'Compare_Segregation'

renanxcortes committed:
removing some extra header lines

renanxcortes committed:
fixing 'hyphotesis' to 'hypothesis' typo

renanxcortes committed:
fixing 'of' typo in docstring

renanxcortes committed:
fixing 'evenness' typo

renanxcortes committed:
removing geometry addition outside of condition (fixed) and adding a hidden _class_name for plotting method

renanxcortes committed:
removing old comments

renanxcortes committed:
add the plot method for Compare_Segregation

renanxcortes committed:
add point estimation value to both inference wrappers

renanxcortes committed:
adding extra condition for spatial indexes that the data input must be a geopandas

renanxcortes committed:
extract only the name of the class with a regex for the inference wrappers

renanxcortes committed:
adding additional conditional for null_approach values

renanxcortes committed:
chaging np.sum to np.nansum due to possible zero value in the data['ti']

renanxcortes committed:
chaging np.sum to np.nansum due to possible zero value in the data['ti']

renanxcortes committed:
adding notebook for paper application

renanxcortes committed:
adding notebook for paper application

renanxcortes committed:
adding segregation paper application notebook

renanxcortes committed:
adding conditional in _return_length_weighted_w function for attaching islands

this commit attach the islands when there are in the data in order to calculate the border length. As a result, the resulting matrix will have the same dimensions of the original data.

renanxcortes committed:
runned notebook with 500 iterations for the segregaiton module paper

renanxcortes committed:
application paper notebook with every simulation runned (including the 500 iterations for systematic approach in Riverside)

ljwolf committed:
fix author email

ljwolf committed:
force propagation of islands through the length-weighted process

renanxcortes committed:
Merge pull request #1 from ljwolf/segregation

length-weighted W

ljwolf committed:
ensure the island dataframe gets built ahead of time and avoid unnecessary unpacking

renanxcortes committed:
putting an extra x

description of this commit

renanxcortes committed:
removing extra x previously put

renanxcortes committed:
Merge pull request #2 from ljwolf/segregation

build islands and avoid unnecessary unpacking.

renanxcortes committed:
include test for dissimilarity function

renanxcortes committed:
putting a different value to check testing feature

renanxcortes committed:
putting back the actual expected value for dissimilarity tester

renanxcortes committed:
adding tests functions

this commit adds several tests functions into the tests repo for each segregation measure  with default input values

renanxcortes committed:
add "inequality." before each script function that rely on other scripts to facilitate the location

renanxcortes committed:
adding the logic of pseudo cumulative distribution comparison inference

renanxcortes committed:
add the cumulative function  in the inference wrapper as a null_approach "pseudo_cumulative_rel"

renanxcortes committed:
relaxing the asumption of having a column named 'geometry' in the geopandas for spatial indexes

This relates to the third point of: https://github.com/renanxcortes/inequality/pull/1#issuecomment-440027958

In addition, this commit remove an unused Y object in the absolute concentration function

renanxcortes committed:
adds the "pseudo_cumulative" option for the null_approach argument

renanxcortes committed:
removing "random_spatial" in the null_approach

this approach used to randomize the spatial topology itself

renanxcortes committed:
removing inequality. from importing function to turn more generalizable later

renanxcortes committed:
add additional step that prevents the result have infinite values or NaN

renanxcortes committed:
adding notebook that have the simulation framework for comparative inference

renanxcortes committed:
adding voronoi lattice generated for concentration measures in comparative segregation scenarios

renanxcortes committed:
removing files from inequality original functions

renanxcortes committed:
updating __init__.py wuth functions

renanxcortes committed:
adding repo with different name

renanxcortes committed:
removing repo with old name

renanxcortes committed:
reorganizing name of the repos

renanxcortes committed:
Merge branch 'segregation' of https://github.com/renanxcortes/inequality into segregation

renanxcortes committed:
chaging the dependencies of the functions to the segregation repo

renanxcortes committed:
updating class_name in inference wrappers to extract only function name in a more reliable way

renanxcortes committed:
update scenarios notebook

works with a global iteration object and a irregular polygon generated by Voronoi concept for the concentration indexes

renanxcortes committed:
Create Scenarios_Single_Index_Allen_2014.ipynb

renanxcortes committed:
fixing objective function to maximize

This address a problem related to the equation 3.4 of Allen et. al. (2014). The objective function wasn't being maximized properly, since only the first term was multiplied by -1. Now it is fixed.

renanxcortes committed:
Update test_density_corrected_dissimilarity.py

renanxcortes committed:
updating the docstring with the function corrected (bjective function)

renanxcortes committed:
Update Scenarios_Single_Index_Allen_2014.ipynb

renanxcortes committed:
swapping parameters orders inside the functions and also adding the 'n' object inside the create_data_frame function which was missing

renanxcortes committed:
adding sample fig

renanxcortes committed:
start of the README.md file

renanxcortes committed:
Update .travis.yml

renanxcortes committed:
Update README.md

renanxcortes committed:
changing the table from latex to markdown in readme

renanxcortes committed:
adding columns separators in the table for markdown in read me

renanxcortes committed:
adding inference table in the readme

renanxcortes committed:
Update README.md

renanxcortes committed:
adding simple example for estimation

renanxcortes committed:
fixing module call brief description and getting started

renanxcortes committed:
merging two sentences in readme

renanxcortes committed:
adding attributes explanation and changing the italic to tt aproach

renanxcortes committed:
removing acronym used in the paper

renanxcortes committed:
adding comparative segregation simuations results (results file and notebooks)

renanxcortes committed:
fixing some identation in docstring

renanxcortes committed:
create profile_wrappers function that calculates point estimation of several measures

currently this function only estimate some specific measures in order to perform a quick segregation profile of the data. Ideally, in the future, would be nice the let the user select what measure he wants to calculate

renanxcortes committed:
fixing description in docstring

renanxcortes committed:
building a general wrapper and fixing all classes that shall be returned

renanxcortes committed:
Create environment.yml

renanxcortes committed:
Update requirements.txt

renanxcortes committed:
Update setup.py

renanxcortes committed:
changing to direct link of libpysal

renanxcortes committed:
removing warnings dependency

renanxcortes committed:
uncommenting to import functions

renanxcortes committed:
adding encoding argument in setup.py (inside the HPCC cluster was not working without this additional argument)

renanxcortes committed:
changing to conda environment installing

renanxcortes committed:
fixing index name in the header of the plotting method

renanxcortes commented:
Closing this PR as this segregation project moved to https://github.com/pysal/segregation.
